### PR TITLE
refactor: merge extract config types

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Merge `ExtractConfig` types and other necessary fields into one interface
- Add the extra fields necessary for extract-cache-fingerprinting to this new interface

## Context
- It will make it simpler if all managers take in same config.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Some bad merge commits in this branch

File: lib/modules/manager/types.ts [hunk 1]
```diff
@@ -15,21 +16,30 @@ export interface ManagerData<T> {
...
   skipInstalls?: boolean;
 }
 
-export interface CustomExtractConfig
-  extends ExtractConfig,
-    RegexManagerTemplates {
+export interface CustomExtractConfig extends RegexManagerTemplates {
   autoReplaceStringTemplate?: string;
-  matchStrings: string[];
+  matchStrings?: string[];
```
Comment (rarkins) [inline lib/modules/manager/types.ts:28]: Isn't this field mandatory?

Comment (RahulGautamSingh) [reply lib/modules/manager/types.ts:28]: It is for regex-manager. But when using the combined config keeping it mandatory causes type issues.

Comment (RahulGautamSingh) [reply lib/modules/manager/types.ts:28]: If we make it mandatory, then we will need to update the tests for each manager except regex Or use a Partial<T> in the test-files. That's a logical thing to do. Will update soon.

File: lib/modules/manager/helmv3/extract.spec.ts [hunk 2]
```diff
@@ -3,6 +3,11 @@ import { DockerDatasource } from '../../datasource/docker';
 import { extractPackageFile } from '.';
 
 jest.mock('../../../util/fs');
+const config = {
```
Comment (viceice) [inline lib/modules/manager/helmv3/extract.spec.ts:6]: please add type when refactor code.

File: lib/modules/manager/regex/strategies.ts [hunk 3]
```diff
@@ -13,8 +13,8 @@ export function handleAny(
   packageFile: string,
   config: CustomExtractConfig
 ): PackageDependency[] {
-  return config.matchStrings
-    .map((matchString) => regEx(matchString, 'g'))
+  return config
+    .matchStrings!.map((matchString) => regEx(matchString, 'g'))
```
Comment (viceice) [inline lib/modules/manager/regex/strategies.ts:17]: please add an override interface to the regex manager where strict required fields are mandatory and validated on initial manager call.

maybe we can also add validation to config validation ( if not already done) and cast to required interface on extractPackageFile call

File: lib/modules/manager/helmv3/extract.spec.ts [hunk 4]
```diff
@@ -1,8 +1,14 @@
...
 import { DockerDatasource } from '../../datasource/docker';
+import type { ExtractConfig } from '../types';
 import { extractPackageFile } from '.';
 
 jest.mock('../../../util/fs');
+const config = {
+  registryAliases: {
+    stable: 'https://charts.helm.sh/stable',
+  },
+} as ExtractConfig;
```
Comment (viceice) [inline lib/modules/manager/helmv3/extract.spec.ts:11]: use partial<T>() helper instead, as cast will hide type issues

File: lib/modules/manager/index.ts [hunk 5]
```diff
@@ -60,7 +60,7 @@ export function extractPackageFile(
   manager: string,
   content: string,
   fileName: string,
-  config: ExtractConfig
+  config?: ExtractConfig
```
Comment (viceice) [inline lib/modules/manager/index.ts:63]: nope, revert. config is mandatory and will always passed

File: lib/modules/manager/regex/index.ts [hunk 6]
```diff
@@ -15,11 +16,15 @@ export const defaultConfig = {
...
+export interface RegexManagerInterface extends CustomExtractConfig {
+  matchStrings: string[];
+}
 export function extractPackageFile(
   content: string,
   packageFile: string,
-  config: CustomExtractConfig
+  extractConfig: ExtractConfig
 ): Result<PackageFile | null> {
+  const config = { ...extractConfig } as RegexManagerInterface;
```
Comment (viceice) [inline lib/modules/manager/regex/index.ts:27]: don't clone

File: lib/modules/manager/regex/types.ts [hunk 7]
```diff
@@ -13,3 +13,7 @@ export interface RecursionParameter {
   index: number;
   combinedGroups: Record<string, string>;
 }
+
+export interface RegexManagerInterface extends CustomExtractConfig {
```
Comment (viceice) [inline lib/modules/manager/regex/types.ts:17]:
```suggestion
export interface RegexManagerConfig extends CustomExtractConfig {
```

**Review by viceice — COMMENTED** [review]
just a small renaming left

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**